concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -150,90 +251,186 @@ exportDatabaseV3 ::
+                                                            -- current genesis index, so use the genesis
+                                                            -- block hash that was previously exported
+                                                            -- for that section and compare it with that
+                                                            -- of the database.
+                                                            _ :|> (gh, _) -> gh
+                                                    if genHash /= exportedGenHash
+                                                        then do
+                                                            liftIO . putStrLn $
+                                                                "Genesis blockhash '"
+                                                                    <> show exportedGenHash
+                                                                    <> "' of most recently exported genesis index "
+                                                                    <> "does not match genesis blockhash '"
+                                                                    <> show genHash
+                                                                    <> "' found in node database for genesis index '"
+                                                                    <> show genIndex
+                                                                    <> "'"
+                                                            return (True, Empty)
+                                                        else do
+                                                            chunks <-
+                                                                writeChunks
+                                                                    genIndex
+                                                                    (demoteProtocolVersion (protocolVersion @pv))
+                                                                    genHash
+                                                                    startHeight
+                                                                    (_bpHeight . sbInfo $ sb)
+                                                                    outDir
+                                                                    chunkSize
+                                                                    overwriteExistingChunks
+                                                            return (False, singleton (genHash, chunks))
+                                        )
+                                        (DBState dbh 0)
                     closeDatabase dbh
-                    exportSections dbDir outDir indexHdl chunkSize (genIndex + 1) 1
-        else putStrLn ("The tree state database does not exist at " ++ treeStateDir)
+                    if exportError
+                        then -- if an error occurred, return
```

Comment:
"return only what we managed to export until now" or some such. It is clear that there is a return.